### DISCUSSION TITLE
MkDocs / Material for MkDocs POC

### DISCUSSION THREAD

**Discussion by martijnschuman** [Windesheim-HBO-ICT/Leerlijn-Wiki#18] [Ideas]
Ik ben gaan kijken naar andere alternatieven en toen kwam ik [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/) tegen, `MFM` als zelfbedachte afkorting 😉. `MFM` is een soort van thema voor `MkDocs`. In theorie zou het beter moeten zijn dan `MkDocs`, in een kleine POC heb ik alleen niet echt voordelen / verschillen kunnen vinden. 

Het grootste nadeel aan `MkDocs` en dus eigenlijk ook aan `MFM` is dat er standaard geen mogelijkheid is om automatisch pagina's via een GitHub repo te laten maken. De pagina's worden gemaakt op basis van de bestanden in de `Docs` map.
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/89603038-f5b3-434c-9832-71ab02d6d98e)

Het is nu gelukt om een docker file met een python script te maken die bij het builden van de applicatie automatisch de git pullt en in mkdocs zet. Ik ga nu kijken of ik de git pull ook bij opstarten van de container kan regelen.

De branch waar ik in werk: [`https://github.com/martijnschuman/Leertaken-wiki/tree/MaterialForMkDocs`](https://github.com/martijnschuman/Leertaken-wiki/tree/MaterialForMkDocs)

## De website na het opstarten van de container
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/9461c99f-33be-43fc-af6e-bef484216dd1)

**Comment by martijnschuman** [comment]
Het is gelukt 😊. Zodra de container opstart wordt het python script uitgevoerd. Dit gebeurt misschien op een onhandige manier, in de `Dockerfile` maak ik een referentie naar een `entrypoint.sh` bestand. Dit `entrypoint.sh` bestand voer twee commando's uit
```
#!/bin/sh
python clone_repo.py
mkdocs serve
```

Hij pult dus eerst de repo, waarna de server pas wordt opgestart.

# Before
(kijk naar navigatiebar)
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/9461c99f-33be-43fc-af6e-bef484216dd1)

# After
Voor de after heb ik eerst in de `Code` branch een nieuwe push gedaan zodat het mapje `JAVA` en twee `.md` bestaden in de repo aanwezig waren. Na het opnieuw opstarten van de container is de pagina `JAVA` met de twee bestanden ook zichtbaar  

![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/8e35895c-85db-4e13-b6ab-8b22d425ee1d)

# Python script
Dit is hoe de python script er nu uit ziet, waarschijnlijk kan dit nog makkelijker of anders maar het werkt wel. Het grootste probleem waar ik mee zat is dat je niet kan clonen naar een map die al bestaat. De bestanden moeten namelijk in een `/app/docs/` map komen te staan. Je zou zeggen van dan verwijder je die map toch voordat je pullt. Ja dat zou kunnen, het probleem is alleen dat in de `/app/docs` map ook de css en andere assets zitten.

Wat ik nu doe is ik clone de map naar `/docs/`. Zodra de clone klaar is worden alle bestanden via `shutil.copy` gekopieerd naar de nieuwe locatie in `/app/docs/`. Het zou misschien nog beter zijn om na het clonen de `/docs/` map all te verwijderen. Dit _durf_ ik alleen niet aan omdat ik het risico niet wil lopen dat het kopieren van een bestand langer duurt, waarna de map al weer wordt verwijderd. Gelijk aan het begin wordt nu gecontroleerd of de `/docs/` map bestaat. Als dat zo is wordt hij verwijderd. Voor verder onderzoek kan het misschien handig zijn om dit te testen met een grote repo of grote bestanden. Enkel om te zien hoe de `Repo.clone_from` en de `shutil.copy` hier mee om gaan.

## De code
```
import os
import shutil
from git import Repo

# function to clone a git repository and copy all files to a destination folder
def clone_repo(repo_url, branch_name, source_folder, destination_folder):
    # Check if the destination directory exists
    if os.path.exists(source_folder):
        # If it does, delete it
        shutil.rmtree(source_folder)

    # clone the repository
    repo = Repo.clone_from(repo_url, source_folder, branch=branch_name)

    # check if repo is cloned
    if(repo):
        # recursively copy all files and folders
        for root, dirs, files in os.walk(source_folder):
            # construct destination folder path
            dest_dir = os.path.join(destination_folder, os.path.relpath(root, source_folder))
            # create destination folder if it doesn't exist
            os.makedirs(dest_dir, exist_ok=True)
            # copy files
            for file_name in files:
                source = os.path.join(root, file_name)
                destination = os.path.join(dest_dir, file_name)

                shutil.copy(source, destination)

clone_repo('https://github.com/martijnschuman/Leertaken-wiki.git', 'Code', '/docs/', '/app/docs/')
```

# ToDo:
1. Dingen zoals de github repo link verplaatsen naar een `.env` of vergelijkbaar bestand
2. Kijken hoe de image met iemand anders gedeeld kan worden op een makkelijke manier
3. Kijken of er een snelkoppeling op het bureaublad te maken is voor het opstarten van de container
4. Kijken of het mogelijk is om de git pull elke interval van bvb 15 minuten of 30 minuten etc uit te voeren is. Ik vermoed dat hiervoor de server eerst gestopt moet worden, aangezien er maar één cmd command tegelijk uitgevoerd kan worden. Misschien dat er iets van `sh` mogelijk is?

**Comment by jaelkwakkel** [comment]
Ziet er goed uit! Voor mij werkt dit alleen als ik de line-endings wijzig van CLRF naar LF. Zou opgelost kunnen worden met een .gitattributes bestand:

https://stackoverflow.com/questions/38905135/why-wont-my-docker-entrypoint-sh-execute 
https://stackoverflow.com/questions/10418975/how-to-change-line-ending-settings

**Comment by martijnschuman** [comment]
Ik heb een nieuwe push gedaan en de bestanden zouden nu standaard `LF` line-endings moeten hebben

**Comment by martijnschuman** [comment]
# Code runner
Het is gelukt om de coderunner te integeren in het project. 
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/776bf793-7ece-4971-b545-71ce57fd8ca6)

Het enige wat in de .md bestanden moet komen is het code blok
`<code-block language="javascript" initialCode='console.log("Hello, world!");'></code-block>`

De coderunner geeft zelf nog een error als ik code wil uitvoeren, maar de JS module kan heel simpel worden ingeladen
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/223e126c-5dc7-4910-9060-938597cbb216)

**Comment by martijnschuman** [comment]
# Tags

Binnen mkdocs is het standaard al mogelijk om tags toe te voegen, een nadeel is alleen dat geneste tags. Dus `For` die onder `C#` zou vallen niet voor gratis gebruikers beschikbaar is. [MKDocs docs](https://squidfunk.github.io/mkdocs-material/setup/setting-up-tags/#nested-tags)

Het is met de gratis vesie ook niet mogelijk om specifieke tags eruit te filteren. Het zou bijvoorbeeld ideaal zijn als je een lijst met alle c# tags kan krijgen. Dan kan deze lijst bijvoorbeeld bovein de introductiepagina komen te staan.

## Optie één
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/ea0c7d52-6c4e-4f5c-aedd-c08901fbe39c)

## Optie twee
![image](https://github.com/Windesheim-HBO-ICT/Leertaken/assets/31829786/bc0e28c8-e327-4392-bfea-c6c2f4e8ad6e)

## Toevoegen
Tags worden nu handmatig toegevoegd door bovenin een .md bestand dit te zetten:
```
---
tags:
  - Tag1
  - Tag2
---
```

Via het python script is het waarschijnlijk wel mogelijk om de tags te laten maken op basis van tags binnen GitHub. Wat ik zosnel online kon vinden kan je `git fetch --tags` doen. Het script zou dan vervolgens elke .md kunnen editen om de tags toe te voegen.

**Comment by martijnschuman** [comment]
# Python script
Ik heb het Python script aangepast zodat de `/app/docs` map voordat er bestanden worden gecloned wordt geleegd op standaard mkdocs bestanden na. Bij de vorige versie was het namelijk zo dat in de git verwijderde mappen en bestanden alsnog bleven bestaan. De bestanden werden enkele gekopieerd en overschreven als ze dubbel waren.

In de nieuwe versie wordt dus alles dat niet standaard met de image meekomt verwijderd, waarna de bestanden opnieuw worden gecloned.
Verder wordt de .git map ook niet meer gekopieerd. Dit gebeurde bij de vorige versie ook wel.

## Het script nu
```
import os
import shutil
from git import Repo

DEFAULT_DOCS_FILES = ['index.md', 'tags.md']
EXCLUDE_DIRS = ['assets', 'stylesheets']


# Remove all files and directories from the destination folder except the default docs files
def clean_destination_folder(destination_folder):
    for root, dirs, files in os.walk(destination_folder):
        for file_name in files:
            if file_name not in DEFAULT_DOCS_FILES and 'assets' not in root and 'stylesheets' not in root:
                os.remove(os.path.join(root, file_name))
        for dir_name in dirs:
            if dir_name not in EXCLUDE_DIRS:
                shutil.rmtree(os.path.join(root, dir_name))

# Copy all files from the source folder to the destination folder
def copy_files(source_folder, destination_folder):
    for root, dirs, files in os.walk(source_folder):
        # Remove the .git directory from the list of directories
        if '.git' in dirs:
            dirs.remove('.git')

        # Create the destination directory if it does not exist
        dest_dir = os.path.join(destination_folder, os.path.relpath(root, source_folder))
        os.makedirs(dest_dir, exist_ok=True)

        # Copy all files from the source folder to the destination folder
        for file_name in files:
            source = os.path.join(root, file_name)
            destination = os.path.join(dest_dir, file_name)
            shutil.copy(source, destination)

# Clone the repository and copy the files to the destination folder
def clone_repo(repo_url, branch_name, source_folder, destination_folder):
    # Remove the source folder if it already exists
    if os.path.exists(source_folder):
        shutil.rmtree(source_folder)

    try:
        # Clone the repository
        repo = Repo.clone_from(repo_url, source_folder, branch=branch_name)

        # Copy the files to the destination folder
        if repo:
            clean_destination_folder(destination_folder)
            copy_files(source_folder, destination_folder)
            print("Repository cloned successfully.")
        else:
            print("Failed to clone repository.")
    except Exception as e:
        print(f"Error: {e}")


clone_repo('https://github.com/martijnschuman/Leertaken-wiki.git', 'Code', '/docs/', '/app/docs/')
```

Als iemand hier feedback op heeft hoor ik het graag
